### PR TITLE
VZ: Support Kernel Image

### PR DESCRIPTION
**Pull request by norio-nomura**
Similar to QEMU's kernel image support.

When using `VZLinuxBootLoader` (https://developer.apple.com/documentation/virtualization/vzlinuxbootloader), the `Virtualization.framework` does not provide EFI to the VM. As a result, the Linux kernel does not recognize a device available as `rtc0`, which means the correct time cannot be obtained during the kernel boot. Consequently, the VM's time is not corrected until `timesync.service` starts via `cloud-init`, which can cause issues like `apt-get` failures. If, for any reason, `timesync.service` fails to correct the time, `apt-get` in the provisioning script might fail, causing `limactl` to perceive the VM boot as failed.

I'm hesitant to add tests for this feature in CI because I'm unsure of the reasons why `timesync.service` might fail to correct the time.

I believe `inject-cmdline-to-template.sh` will be helpful when testing this feature.

### CODE REVIEW

File: hack/inject-cmdline-to-template.sh [hunk 1]
```diff
@@ -9,7 +9,7 @@
 # 6. inject the kernel and initrd location, digest, and cmdline to the template
 # 7. output kernel_location, kernel_digest, cmdline, initrd_location, initrd_digest
 
-set -eu -o pipefail
+set -u -o pipefail
```
Comment (AkihiroSuda) [inline hack/inject-cmdline-to-template.sh:12]: ?

File: pkg/vz/vm_darwin.go [hunk 2]
```diff
@@ -686,6 +681,40 @@ func getMachineIdentifier(driver *driver.BaseDriver) (*vz.GenericMachineIdentifi
...
+
+	efiVariableStore, err := getEFI(driver)
+	if err != nil {
+		return nil, err
+	}
+	logrus.Debugf("Using EFI Boot Loader")
+	return vz.NewEFIBootLoader(vz.WithEFIVariableStore(efiVariableStore))
+}
+
+func linuxBootLoader(driver *driver.BaseDriver) (*vz.LinuxBootLoader, error) {
```
Comment (AkihiroSuda) [inline pkg/vz/vm_darwin.go:697]: Please add a comment to clarify that this function may return nil without an error

Comment (norio-nomura) [reply pkg/vz/vm_darwin.go:697]: changed `linuxBootLoader()` to return error, then explicitly ignores the error on caller.

File: hack/inject-cmdline-to-template.sh (hunk 1 above)
Comment (norio-nomura) [reply hack/inject-cmdline-to-template.sh]: changed to use `-e` again

**Comment by norio-nomura** [comment]
When using this change to boot a VM with a kernel image, `lima-guestagent` will continuously throw errors because `/dev/rtc` is not present in the VM.
```console
$ limactl shell docker sudo systemctl status lima-guestagent
● lima-guestagent.service - lima-guestagent
     Loaded: loaded (/etc/systemd/system/lima-guestagent.service; enabled; preset: enabled)
     Active: active (running) since Thu 2024-08-29 08:52:54 JST; 10min ago
   Main PID: 1670 (lima-guestagent)
      Tasks: 7 (limit: 19101)
     Memory: 7.1M (peak: 8.3M)
        CPU: 2.010s
     CGroup: /system.slice/lima-guestagent.service
             └─1670 /usr/local/bin/lima-guestagent daemon --vsock-port 2222

Aug 29 09:02:14 lima-docker lima-guestagent[1670]: time="2024-08-29T09:02:14+09:00" level=warning msg="fixSystemTimeSkew: lookup error: open /dev/rtc: no such file or directory"
Aug 29 09:02:24 lima-docker lima-guestagent[1670]: time="2024-08-29T09:02:24+09:00" level=warning msg="fixSystemTimeSkew: lookup error: open /dev/rtc: no such file or directory"
Aug 29 09:02:34 lima-docker lima-guestagent[1670]: time="2024-08-29T09:02:34+09:00" level=warning msg="fixSystemTimeSkew: lookup error: open /dev/rtc: no such file or directory"
Aug 29 09:02:44 lima-docker lima-guestagent[1670]: time="2024-08-29T09:02:44+09:00" level=warning msg="fixSystemTimeSkew: lookup error: open /dev/rtc: no such file or directory"
Aug 29 09:02:54 lima-docker lima-guestagent[1670]: time="2024-08-29T09:02:54+09:00" level=warning msg="fixSystemTimeSkew: lookup error: open /dev/rtc: no such file or directory"
Aug 29 09:03:04 lima-docker lima-guestagent[1670]: time="2024-08-29T09:03:04+09:00" level=warning msg="fixSystemTimeSkew: lookup error: open /dev/rtc: no such file or directory"
Aug 29 09:03:14 lima-docker lima-guestagent[1670]: time="2024-08-29T09:03:14+09:00" level=warning msg="fixSystemTimeSkew: lookup error: open /dev/rtc: no such file or directory"
Aug 29 09:03:24 lima-docker lima-guestagent[1670]: time="2024-08-29T09:03:24+09:00" level=warning msg="fixSystemTimeSkew: lookup error: open /dev/rtc: no such file or directory"
Aug 29 09:03:34 lima-docker lima-guestagent[1670]: time="2024-08-29T09:03:34+09:00" level=warning msg="fixSystemTimeSkew: lookup error: open /dev/rtc: no such file or directory"
Aug 29 09:03:44 lima-docker lima-guestagent[1670]: time="2024-08-29T09:03:44+09:00" level=warning msg="fixSystemTimeSkew: lookup error: open /dev/rtc: no such file or directory"
```

File: pkg/vz/vm_darwin.go [hunk 3]
```diff
@@ -686,6 +681,45 @@ func getMachineIdentifier(driver *driver.BaseDriver) (*vz.GenericMachineIdentifi
...
+	kernel := filepath.Join(driver.Instance.Dir, filenames.Kernel)
+	kernelCmdline := filepath.Join(driver.Instance.Dir, filenames.KernelCmdline)
+	initrd := filepath.Join(driver.Instance.Dir, filenames.Initrd)
+	if _, err := os.Stat(kernel); err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			logrus.Debugf("Kernel file %q not found", kernel)
+		} else {
+			logrus.Debugf("Error while checking kernel file %q: %v", kernel, err)
```
Comment (AkihiroSuda) [inline pkg/vz/vm_darwin.go:706]:
```suggestion
			logrus.WithError(err).Debugf("Error while checking kernel file %q", kernel)
```

File: pkg/vz/vm_darwin.go [hunk 4]
```diff
@@ -686,6 +681,41 @@ func getMachineIdentifier(driver *driver.BaseDriver) (*vz.GenericMachineIdentifi
 	return vz.NewGenericMachineIdentifierWithDataPath(identifier)
 }
 
+func bootLoader(driver *driver.BaseDriver) (vz.BootLoader, error) {
+	// ignore the error it might be a non-existing kernel file
+	if linuxBootLoder, _ := linuxBootLoader(driver); linuxBootLoder != nil {
```
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply pkg/vz/vm_darwin.go:686]: @norio-nomura WDYT?

[the remark this answers is not in the document]
Comment (norio-nomura) [reply pkg/vz/vm_darwin.go]: Updated!

**Comment by balajiv113** [comment]
@norio-nomura Do you have a sample yaml for running with kernel ??
Can you please share it.

**Comment by AkihiroSuda** [comment]
Also could you squash the commits for `pkg/vz/*.go`?

**Comment by norio-nomura** [comment]
> @norio-nomura Do you have a sample yaml for running with kernel ?? Can you please share it.

You can inject kernel related info into `ubuntu.yaml` by following:
```console
hack/inject-cmdline-to-template.sh templates/ubuntu.yaml console=hvc0
```

e.g.
```console
$ hack/inject-cmdline-to-template.sh templates/ubuntu.yaml console=hvc0
kernel_location=https://cloud-images.ubuntu.com/releases/24.04/release-20240821/unpacked/ubuntu-24.04-server-cloudimg-arm64-vmlinuz-generic
kernel_digest=sha256:69e66230d6acc503082f098fd072a672a804774007209110b353d25bb7007669
cmdline=root=LABEL=cloudimg-rootfs ro console=tty1 console=ttyAMA0 console=hvc0
initrd_location=https://cloud-images.ubuntu.com/releases/24.04/release-20240821/unpacked/ubuntu-24.04-server-cloudimg-arm64-initrd-generic
initrd_digest=sha256:92acf9f2976b7a8b3b5e9e3f22211d17a6fdb3f2aed981467f91cd6dbfa16ae5
$ limactl start templates/ubuntu.yaml --tty=false
INFO[0000] Terminal is not available, proceeding without opening an editor 
INFO[0000] Starting the instance "ubuntu" with VM driver "vz" 
WARN[0000] vmType vz: ignoring images[1]: [Kernel Initrd] 
INFO[0000] Attempting to download the image              arch=aarch64 digest="sha256:5ecac6447be66a164626744a87a27fd4e6c6606dc683e0a233870af63df4276a" location="https://cloud-images.ubuntu.com/releases/24.04/release-20240821/ubuntu-24.04-server-cloudimg-arm64.img"
INFO[0000] Using cache "/Users/norio/Library/Caches/lima/download/by-url-sha256/346ee1ff9e381b78ba08e2a29445960b5cd31c51f896fc346b82e26e345a5b9a/data" 
INFO[0000] Attempting to download the kernel             arch=aarch64 digest="sha256:69e66230d6acc503082f098fd072a672a804774007209110b353d25bb7007669" location="https://cloud-images.ubuntu.com/releases/24.04/release-20240821/unpacked/ubuntu-24.04-server-cloudimg-arm64-vmlinuz-generic"
INFO[0000] decompressing .04-server-cloudimg-arm64-vmlinuz-generic with gzip 
INFO[0000] Decompressing data                           
17.39 MiB / 17.39 MiB [------------------------------------------] 100.00% ? p/s
INFO[0000] Using cache "/Users/norio/Library/Caches/lima/download/by-url-sha256/98c212aa1b30c23b56134728fa6adf3096bea13b93cf1744bf88d35d488be2e7/data" 
INFO[0000] Attempting to download the initrd             arch=aarch64 digest="sha256:92acf9f2976b7a8b3b5e9e3f22211d17a6fdb3f2aed981467f91cd6dbfa16ae5" location="https://cloud-images.ubuntu.com/releases/24.04/release-20240821/unpacked/ubuntu-24.04-server-cloudimg-arm64-initrd-generic"
INFO[0000] Using cache "/Users/norio/Library/Caches/lima/download/by-url-sha256/c972a5d4d42628ec1dd6610a5398674a7f85a3bfa916d614064809722e0f4840/data" 
INFO[0000] Converting "/Users/norio/.lima/ubuntu/basedisk" (qcow2) to a raw disk "/Users/norio/.lima/ubuntu/diffdisk" 
3.50 GiB / 3.50 GiB [-------------------------------------] 100.00% 213.27 MiB/s
INFO[0017] Expanding to 100GiB                          
INFO[0017] Attempting to download the nerdctl archive    arch=aarch64 digest="sha256:77c747f09853ee3d229d77e8de0dd3c85622537d82be57433dc1fca4493bab95" location="https://github.com/containerd/nerdctl/releases/download/v1.7.6/nerdctl-full-1.7.6-linux-arm64.tar.gz"
INFO[0017] Using cache "/Users/norio/Library/Caches/lima/download/by-url-sha256/21cc8dfa548ea8a678135bd6984c9feb9f8a01901d10b11bb491f6f4e7537158/data" 
WARN[0018] [hostagent] GRPC port forwarding is experimental 
INFO[0018] [hostagent] hostagent socket created at /Users/norio/.lima/ubuntu/ha.sock 
INFO[0018] [hostagent] Starting VZ (hint: to watch the boot progress, see "/Users/norio/.lima/ubuntu/serial*.log") 
INFO[0018] [hostagent] new connection from  to          
INFO[0018] SSH Local Port: 59900                        
INFO[0018] [hostagent] Waiting for the essential requirement 1 of 2: "ssh" 
INFO[0018] [hostagent] [VZ] - vm state change: running  
INFO[0028] [hostagent] Waiting for the essential requirement 1 of 2: "ssh" 
INFO[0028] [hostagent] The essential requirement 1 of 2 is satisfied 
INFO[0028] [hostagent] Waiting for the essential requirement 2 of 2: "user session is ready for ssh" 
INFO[0028] [hostagent] The essential requirement 2 of 2 is satisfied 
INFO[0028] [hostagent] Waiting for the optional requirement 1 of 2: "systemd must be available" 
INFO[0028] [hostagent] Guest agent is running           
INFO[0028] [hostagent] The optional requirement 1 of 2 is satisfied 
INFO[0028] [hostagent] Not forwarding TCP 127.0.0.54:53 
INFO[0028] [hostagent] Not forwarding TCP 127.0.0.53:53 
INFO[0028] [hostagent] Waiting for the optional requirement 2 of 2: "containerd binaries to be installed" 
INFO[0028] [hostagent] Not forwarding TCP [::]:22       
INFO[0028] [hostagent] Not forwarding UDP 127.0.0.54:53 
INFO[0028] [hostagent] Not forwarding UDP 127.0.0.53:53 
INFO[0028] [hostagent] Not forwarding UDP 192.168.5.15:68 
INFO[0037] [hostagent] The optional requirement 2 of 2 is satisfied 
INFO[0037] [hostagent] Waiting for the guest agent to be running 
INFO[0037] [hostagent] Waiting for the final requirement 1 of 1: "boot scripts must have finished" 
INFO[0037] [hostagent] The final requirement 1 of 1 is satisfied 
INFO[0038] READY. Run `limactl shell ubuntu` to open the shell. 
```

**Review by balajiv113 — APPROVED** [review]
LGTM 👍 

Please squash the commits

**Comment by norio-nomura** [comment]
Thanks! 🙏🏻

**Pull request merged**